### PR TITLE
PICARD-2101: Enhance script autocompletion with dynamic variable parsing and plugin support

### PR DESCRIPTION
**Pull request by knguyen1**
# Summary

* This is a…
  * [ ] Bug fix
  * [x] Feature addition
  * [ ] Refactoring
  * [ ] Minor / simple change (like a typo)
  * [ ] Other
* **Describe this change in 1-2 sentences**: Enhanced script autocompletion with dynamic variable parsing, plugin variable support, and context-aware completion for improved user experience.

# Problem

* JIRA ticket (_optional_): PICARD-2101

The current script autocompletion system has several limitations:
- Missing custom variables: When users create variables with `$set()`, those variables don't appear in autocompletion
- No dynamic discovery: The system doesn't learn from existing scripts or current script content  
- Limited context awareness: No smart suggestions based on what the user is currently working on
- No plugin extensibility: Plugins cannot register additional variables for completion

# Solution

Implemented three major enhancements to the script completion system:

**1. Dynamic Script Parsing**
- Real-time parsing of currently edited script to extract `$set(variable_name, ...)` calls
- Uses existing ScriptParser for safe parsing with AST traversal
- Caches parsed variables and updates on script changes
- Includes regex fallback for resilience during live edits

**2. Plugin Variable Extension** 
- New extension point `ext_point_script_variables` for plugins to register variables
- `register_script_variable(name, documentation)` function for plugin integration
- Plugin variables appear in completion with `%variable_name%` format
- Full integration with existing completion ranking and sorting

**3. Context-Aware Completion**
- Smart completion based on cursor position and context
- Function parameter completion: Shows tag names when cursor is inside `$set(`, `$get(`, etc.
- Variable usage completion: Shows relevant variables when typing `%`
- Smart filtering: Prioritizes recently used or contextually relevant variables
- Handles complex nested structures and edge cases

**Technical Implementation:**
- Extended `ScriptCompleter` with dynamic variable extraction methods
- Added comprehensive test coverage for all new functionality
- Maintains backward compatibility with existing completion behavior
- Follows established extension point patterns used throughout Picard

# Action

Additional actions required:
* [ ] Update Picard [documentation](https://github.com/metabrainz/picard-docs) (please include a reference to this PR)
* [ ] Other (please specify below)

### CODE REVIEW

File: picard/ui/widgets/scripttextedit.py [hunk 1]
```diff
@@ -173,21 +180,109 @@ def highlightBlock(self, text):
...
+            candidates = list(builtin_vars | user_vars | plugin_vars)
+            candidates.sort(key=lambda x: (-self._var_usage_counts.get(x, 0), x))
+            for name in candidates:
+                yield name
+            return
+
+        # Default: functions then variables, variables ranked by usage count
+        if mode in (CompletionMode.DEFAULT, CompletionMode.FUNCTION_NAME):
+            for name in sorted(script_function_names()):
+                yield f'${name}'
```
Comment (zas) [inline picard/ui/widgets/scripttextedit.py]: I think it can be simplified:


```python
        # Default: functions then variables, variables ranked by usage count
        if mode in (CompletionMode.DEFAULT, CompletionMode.FUNCTION_NAME):
            for name in sorted(script_function_names()):
                yield f'${name}'

        builtin_vars = set(script_variable_tag_names())
        plugin_vars = get_plugin_variable_names()
        user_vars = set(v for v in self._user_defined_variables if v not in builtin_vars and v not in plugin_vars)
        all_vars = list(builtin_vars | user_vars | plugin_vars)
        all_vars.sort(key=lambda x: (-self._var_usage_counts.get(x, 0), x))

        if mode == CompletionMode.TAG_NAME_ARG and context.get('arg_index', 0) == 0:
            # Suggest bare tag names for functions expecting a tag name as first argument.
            # Order by usage count (descending), then alphabetically.
            yield name from all_vars
            return

        if mode in (CompletionMode.DEFAULT, CompletionMode.VARIABLE):
            for name in all_vars:
                yield f'%{name}%'
```

File: picard/ui/widgets/scripttextedit.py [hunk 2]
```diff
@@ -173,21 +180,109 @@ def highlightBlock(self, text):
...
+    def _count_variable_usage(self, script_content: str) -> dict[str, int]:
+        """Count variable usages in the script content."""
+        counts: dict[str, int] = {}
+        for m in re.finditer(r'%([A-Za-z0-9_:]+)%', script_content):
+            name = m.group(1)
+            counts[name] = counts.get(name, 0) + 1
+        # Also cout references via `$get(name)`
+        for m in re.finditer(r'\$get\(\s*([A-Za-z0-9_]+)\s*\)', script_content):
+            name = m.group(1)
+            counts[name] = counts.get(name, 0) + 1
```
Comment (zas) [inline picard/ui/widgets/scripttextedit.py]: Code redundancy can be reduced here:

```python
        for regexp in (r'%([A-Za-z0-9_:]+)%', '\$get\(\s*([A-Za-z0-9_]+)\s*\)'):
            for m in re.finditer(regexp, script_content):
                name = m.group(1)
                counts[name] = counts.get(name, 0) + 1
```

File: picard/ui/widgets/variable_extractor.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                continue
+            with contextlib.suppress(ScriptError):
+                expression = self._parser.parse(line)
+                self._collect_from_ast(expression, names)
+        return names
+
+    def _collect_from_regex(self, script_content: str) -> set[str]:
+        """Collect variable names from a regex pattern of the script content."""
+        return {
+            m.group(1) for m in re.finditer(r"\$set\(\s*([A-Za-z0-9_\u00C0-\u017F\u4E00-\u9FFF]+)\s*,", script_content)
```
Comment (Sophist-UK) [inline picard/ui/widgets/variable_extractor.py]: Would it be better to pre-compile this regex once at Picard initialisation?

Comment (knguyen1) [reply picard/ui/widgets/variable_extractor.py]: [refactor(variable_extractor): optimize regex pattern for variable ext…](https://github.com/metabrainz/picard/pull/2742/commits/14c5dfeed9eb4337dc0cc9a847985f18f46f634e)

File: picard/ui/widgets/scripttextedit.py (hunk 2 above)
Comment (knguyen1) [reply picard/ui/widgets/scripttextedit.py]: [refactor(scripttextedit): optimize variable usage counting with pre-c…](https://github.com/metabrainz/picard/pull/2742/commits/cbb59afdadceeb04ce285313675839790e8b6cad)

File: picard/ui/widgets/scripttextedit.py (hunk 1 above)
Comment (knguyen1) [reply picard/ui/widgets/scripttextedit.py]: [refactor(scripttextedit): streamline variable suggestion logic in Scr…](https://github.com/metabrainz/picard/pull/2742/commits/1a6528caea4027750a611cee8c8dc2f63210c10b)

File: picard/extension_points/script_variables.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    Returns
+    -------
+    set[str]
+        Set of variable names provided by plugins
+    """
+    variables = set()
+    for name, _ in ext_point_script_variables:
+        variables.add(name)
+    return variables
```
Comment (zas) [inline picard/extension_points/script_variables.py]: Please do not use `_` for unused variables since it is used by gettext.
Also maybe make it one line:

```python
return {name for name, __unused in ext_point_script_variables}
```

Comment (knguyen1) [reply picard/extension_points/script_variables.py]: [refactor(script_variables): simplify variable name retrieval in get_p…](https://github.com/metabrainz/picard/pull/2742/commits/5a9eccc05ac5a778e3b53ca9f1cd2f0b56501455)

File: picard/ui/widgets/completion_provider.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        The method sorts variables by usage count (descending) and then
+        alphabetically. Different completion modes return different formats:
+        - DEFAULT/FUNCTION_NAME: Returns function names with $ prefix
+        - TAG_NAME_ARG: Returns variable names without formatting
+        - DEFAULT/VARIABLE: Returns variable names with % prefix and suffix
+        """
+        plugin_variables = self._get_plugin_variable_names()
+        builtin_variables = set(builtin_variables)
+        user_variables = set(
+            v for v in user_defined_variables if v not in builtin_variables and v not in plugin_variables
```
Comment (zas) [inline picard/ui/widgets/completion_provider.py]: Just a minor thing, but since we union all sets, (the set) `user_defined_variables` can replace `user_variables` totally (it is only used there).

```python
plugin_variables = self._get_plugin_variable_names()
builtin_variables = set(builtin_variables)
user_variables = set(
    v for v in user_defined_variables if v not in builtin_variables and v not in plugin_variables
)
all_variables = list(builtin_variables | user_variables | plugin_variables)
```

becomes:

```python
plugin_variables = self._get_plugin_variable_names()
builtin_variables = set(builtin_variables)
all_variables = list(builtin_variables | user_defined_variables | plugin_variables)
```

**Review by zas — CHANGES_REQUESTED** [review]

File: picard/ui/widgets/completion_provider.py (hunk 5 above)
Comment (knguyen1) [reply picard/ui/widgets/completion_provider.py]: Good catch. [refactor(completion_provider): simplify variable collection logic in …](https://github.com/metabrainz/picard/pull/2742/commits/6096dd884c936d41aa635d2dfb19fd862d024d1c)

File: picard/ui/widgets/user_script_scanner.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        Returns
+        -------
+        set[str]
+            Set of variable names found in all user scripts.
+        """
+        all_variables = set()
+
+        with contextlib.suppress(AttributeError, TypeError, ValueError):
+            current_scripts = list(iter_active_tagging_scripts())
+            current_hash = hash(tuple((s.name, s.content) for s in current_scripts))
```
Comment (zas) [inline picard/ui/widgets/user_script_scanner.py]: Do we really need to hash on `s.name` here?
I mean, below, we parse only content for variables, so if only name changed the result should be the same.

Also, `current_scripts` can be a `tuple`.
And those 2 lines can be moved to a function as they are used twice.

Comment (knguyen1) [reply picard/ui/widgets/user_script_scanner.py]: [refactor(user_script_scanner): extract script hash calculation into a…](https://github.com/metabrainz/picard/pull/2742/commits/3af1e8f75c2a449aac67a4278485bb6e3b80887d)

I think name should be kept. Some possible bugs if we remove name...

1. Error logging inconsistency - logs show new names but autocompletion uses old context
2. UI state synchronization - different parts of the UI get out of sync
3. Script execution context confusion - duplicate names not detected
4. User experience inconsistency - what user sees vs what system knows

These bugs would create a fragmented user experience where different parts of the application show different information about the same scripts.

**Comment by rdswift** [comment]
Will it properly interpret cases where the variable name is passed indirectly to the `$set()` command?  For example:

```taggerscript
$noop( Create a separate tag for each artist listed for the
       track as 'artist_1', 'artist_2', etc. )
$foreach(%artists%,$set(artist_%_loop_count%,%_loop_value%))
```

Will this include the resulting `artist_1`, `artist_2`, etc. variables in the autocomplete list?

**Comment by rdswift** [comment]
Something for future consideration is to allow translated description strings for variables added (registered) by a plugin.  I suppose that could be handled within the new plugin system, so the translated description is passed when registering the variable.

**Comment by knguyen1** [comment]
>Will this include the resulting artist_1, artist_2, etc. variables in the autocomplete list?

Out of scope of original JIRA requirement.  Current AST parsing + regex fallback only covers statically named variables.

~~If you'd like me to proceed with dynamic var's, have to vote on it.  I think it can be done with `variable_extractor.py` changes.~~

EDIT: Way too complex to solve--I will not be doing it.  Besides, I think your example `$foreach(%artists%,$set(artist_%_loop_count%,%_loop_value%))` relies on runtime context (the `%artists%` value) that the script parser does not have at time of parsing.  One can probably write a parser for known formats like `$foreach(1;2;3,$set(test_%_loop_count%,%_loop_value%))` though.

**Review by zas — APPROVED** [review]
LGTM

Tested, work as expected for me.

File: picard/script/variable_pattern.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import re
+
+
+# Character class for variable names: Unicode letters/digits ("\w") and colon.
+# Python's regex "\w" includes Unicode letters, digits, and underscores.
+VAR_NAME_CLASS = r"[\w:]"
+_VARIABLE_NAME_GROUP = rf"({VAR_NAME_CLASS}+)"
+
+
+def pattern_percent_variable() -> re.Pattern:
```
Comment (phw) [inline picard/script/variable_pattern.py]: Why wrap the patterns into functions, and not directly declare them as constants? They are static and don't change. A function would be useful if they would need to be generated dynamically.

```python
PATTER_PERCENT_VARIABLE = re.compile(rf"%({_VARIABLE_NAME_GROUP})%")
```

**Comment by phw** [comment]
> EDIT: Way too complex to solve--I will not be doing it. Besides, I think your example `$foreach(%artists%,$set(artist_%_loop_count%,%_loop_value%))` relies on runtime context (the `%artists%` value) that the script parser does not have at time of parsing. One can probably write a parser for known formats like `$foreach(1;2;3,$set(test_%_loop_count%,%_loop_value%))` though.

Agreed that this is too complex to solve, and won't capture any case that requires context. Let's focus on statically declared variable names.

File: picard/script/variable_pattern.py (hunk 7 above)
Comment (knguyen1) [reply picard/script/variable_pattern.py]: @phw Good call. [refactor(variable_patterns): consolidate regex pattern definitions an…](https://github.com/metabrainz/picard/pull/2742/commits/df21bd7b72ff6dc0a0b7eae724f5f9124cf14f6f)

**Comment by knguyen1** [comment]
@phw That's out of scope of this PR's changes; it touches the file `picard/ui/options/scripting.py`.  Regardless here is the change needed: [fix(scripting_options): bug red error bar + add debounce](https://github.com/metabrainz/picard/pull/2742/commits/1ffd00b0f69d6e3a40a6fba6270e82791b25f12e)

Also, I found it strange that the red error bar is unforgiving against pauses in typing.  So added a debounce of 350ms.

**Comment by phw** [comment]
> That's out of scope of this PR's changes; it touches the file `picard/ui/options/scripting.py`. Regardless here is the change needed: [fix(scripting_options): bug red error bar + add debounce](https://github.com/metabrainz/picard/pull/2742/commits/1ffd00b0f69d6e3a40a6fba6270e82791b25f12e)

You are actually right, same issue is on master branch. So that's a regression on master already (works fine in the 2.x versions). I had assumed it would be broken in this branch only. But good to have it fixed regardless.
